dag-json/src/codec.rs

```diff
@@ -168,15 +177,36 @@ impl<'de> de::Visitor<'de> for JsonVisitor {
             values.push((key, value));
         }
 
-        // JSON Object represents IPLD Link if it is `{ "/": "...." }` therefor
-        // we valiadet if that is the case here.
+        // JSON Object represents an IPLD Link if it is a slash, followed by a string
+        // (`{ "/": "...." }`) therefore we validate if that is the case here.
         if let Some((key, WrapperOwned(Ipld::String(value)))) = values.first() {
-            if key == LINK_KEY && values.len() == 1 {
-                let link = base64::decode(value).map_err(SerdeError::custom)?;
-                let cid = Cid::try_from(link).map_err(SerdeError::custom)?;
+            if key == RESERVED_KEY && values.len() == 1 {
+                let cid = Cid::try_from(value.clone()).map_err(SerdeError::custom)?;
                 return Ok(Ipld::Link(cid));
             }
         }
+        // JSON Object represents IPLD bytes if it is a slash, followed by an object which contains
+        // only a single key called "bytes", where the value is a string.
+        if let Some((key, WrapperOwned(Ipld::Map(map)))) = values.first() {
+            // Object with a slash
+            if key == RESERVED_KEY && values.len() == 1 {
+                if let Some((bytes_key, Ipld::String(bytes_value))) = map.iter().next() {
+                    if bytes_key == BYTES_KEY && values.len() == 1 {
+                        let multibase_prefixed = Base::Base64.code().to_string() + bytes_value;
+                        if let Ok((decoded_base, decoded_bytes)) =
+                            multibase::decode(multibase_prefixed)
+                        {
+                            if decoded_base != Base::Base64 {
```

Comment:
I don't understand your suggestion. The `bytes_value` is a base64 encoded string, without a multibase prefix. That can't be decoded by a multibase implementation.

I'm use multibase and not base64 directly as I didn't want to have another depdendency and also wanted to be sure we use a compatible base64 encoding. I just forgot to remove the `base64` dependency from the crate. Thanks for the pointing me to it, I'll do that now.

Reply:
Sorry I wasn't being clear, let me try again:

It seems to me we have this base64 string `bytes_value` and we're trying to decode it. To do that we are currently adding a string prefix to the string, this results in copying `bytes_value` into a new string. I'm saying we don't have to do that and we don't need the base64 crate, we can do it with multibase directly. Since each base in multibase has a `decode` [method](https://docs.rs/multibase/0.9.1/multibase/enum.Base.html#method.decode).

So the diff would look something like this:

```
diff --git a/dag-json/src/codec.rs b/dag-json/src/codec.rs
index 16c36bd..365db1e 100644
--- a/dag-json/src/codec.rs
+++ b/dag-json/src/codec.rs
@@ -192,17 +192,10 @@ impl<'de> de::Visitor<'de> for JsonVisitor {
             if key == RESERVED_KEY && values.len() == 1 {
                 if let Some((bytes_key, Ipld::String(bytes_value))) = map.iter().next() {
                     if bytes_key == BYTES_KEY && values.len() == 1 {
-                        let multibase_prefixed = Base::Base64.code().to_string() + bytes_value;
-                        if let Ok((decoded_base, decoded_bytes)) =
-                            multibase::decode(multibase_prefixed)
-                        {
-                            if decoded_base != Base::Base64 {
-                                return Err(SerdeError::custom(
-                                    "bytes kind must be base-64 encoded",
-                                ));
-                            }
-                            return Ok(Ipld::Bytes(decoded_bytes));
-                        }
+                        let decoded_bytes = Base::Base64
+                            .decode(bytes_value)
+                            .map_err(SerdeError::custom)?;
+                        return Ok(Ipld::Bytes(decoded_bytes));
                     }
                 }
             }
```

Note that `Base::Base64` comes from multibase.

Reply:
I'm sorry. I thought `Base::Base64.decode()` would also expect a prefixed string. I didn't look up the API. Thanks for bringing it up, this is indeed a good idea.